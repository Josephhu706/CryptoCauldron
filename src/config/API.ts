import axios from "axios";

export type Coin = {
    id: string;
    symbol: string;
    name: string;
    image: string;
    current_price: number;
    market_cap: number;
    market_cap_rank: number;
    fully_diluted_valuation: number;
    total_volume: number;
    high_24h: number;
    low_24h: number;
    price_change_24h: number;
    price_change_percentage_24h: number;
    market_cap_change_24h: number;
    market_cap_change_percentage_24h: number;
    circulating_supply: number;
    total_supply: number;
    max_supply: number;
    ath: number;
    ath_change_percentage: number;
    ath_date: string;
    atl: number;
    atl_change_percentage: number;
    atl_date: string;
    roi: any;
    last_updated: string;
}

export const CoinList = () =>
`https://api.coingecko.com/api/v3/coins/markets?vs_currency=aud&order=market_cap_desc&per_page=100&page=1&sparkline=false`;

export const SingleCoin = (id:number) =>
`https://api.coingecko.com/api/v3/coins/${id}`;

export const HistoricalChart = (id:number, days = 365) =>
`https://api.coingecko.com/api/v3/coins/${id}/market_chart?vs_currency=aud&days=${days}`;

export const TrendingCoins = () =>
`https://api.coingecko.com/api/v3/coins/markets?vs_currency=aud&order=gecko_desc&per_page=10&page=1&sparkline=false&price_change_percentage=24h`;

export const fetchCrypto = async ()=>{
    //put page not found 
    const endpoint = `https://api.coingecko.com/api/v3/coins/markets?vs_currency=aud&order=market_cap_desc&per_page=100&page=1&sparkline=false`;
    const {data} = await axios.get(endpoint)
    console.log(data)
    return data

}