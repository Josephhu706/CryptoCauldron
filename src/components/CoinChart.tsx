import React, { useEffect, useState } from "react";
import { fetchHistory, SingleCoin } from "../config/API";
import CircularProgress from "@mui/material/CircularProgress";
import callError from "../assets/images/call-error.svg";
import { ThemeProvider, createTheme } from "@mui/material/styles";
import { CssBaseline } from "@mui/material/";
import "../styles/CoinChartStyles.scss";
import { Line } from "react-chartjs-2";
import Button from "@mui/material/Button";

import {
  Chart as ChartJS,
  CategoryScale,
  LinearScale,
  PointElement,
  LineElement,
  Title,
  Tooltip,
  Legend,
} from "chart.js";

ChartJS.register(
  CategoryScale,
  LinearScale,
  PointElement,
  LineElement,
  Title,
  Tooltip,
  Legend
);

const darkTheme = createTheme({
  palette: {
    mode: "dark",
  },
});

type Props = {
  coin: any;
};

const CoinChart: React.FC<Props> = ({ coin }) => {
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<string>("");
  const [history, setHistory] = useState<any>([]);
  const [days, setDays] = useState<number>(1);
  const [lineData, setLineData] = useState([]);
  const chartTimelines = [
    {
      label: "24 Hours",
      value: 1,
    },
    {
      label: "30 Days",
      value: 30,
    },
    {
      label: "3 Months",
      value: 90,
    },
    {
      label: "1 Year",
      value: 365,
    },
  ];

  useEffect(() => {
    getHistory(coin, days);
    // eslint-disable-next-line
  }, []);

  const getHistory = async (id: string, days: number) => {
    setLoading(true);
    try {
      const newCoin = await fetchHistory(id, days);
      setHistory(newCoin);

      let newLineData = newCoin.map((coin: Array<number>) => {
        let date = new Date(coin[0]);
        let time =
          date.getHours() > 12
            ? `${date.getHours() - 12}:${date.getMinutes()} PM`
            : `${date.getHours()}:${date.getMinutes()} AM`;
        return days === 1 ? time : date.toLocaleDateString();
      });
      setLineData(newLineData);
    } catch (error: any) {
      setError(error);
    }
    setLoading(false);
  };

  const createHistory = (value: number) => {
    setDays(value);
    createLine();
  };

  const createLine = async () => {
    let newLineData = history.map((coin: Array<number>) => {
      let date = new Date(coin[0]);
      let time =
        date.getHours() > 12
          ? `${date.getHours() - 12}:${date.getMinutes()} PM`
          : `${date.getHours()}:${date.getMinutes()} AM`;
      return days === 1 ? time : date.toLocaleDateString();
    });
    setLineData(newLineData);
  };

  return (
    <>
      <ThemeProvider theme={darkTheme}>
        <CssBaseline />
        <div className="container">
          <h1 className="coinHeading">Coin Price Data</h1>
          {loading ? (
            <div className="loadingProgressContent">
              <CircularProgress
                className="loadingProgress"
                data-testid="loading"
              />
            </div>
          ) : error ? (
            <div data-testid="fetchError" className="fetchError">
              <img
                src={callError}
                style={{ height: "20em" }}
                alt="fetchError"
              />
            </div>
          ) : (
            <>
              <div data-testid="coinChart" className="coinChart">
                <Line
                  data={{
                    labels: lineData,
                    datasets: [
                      {
                        data: history?.map((coin: any) => coin[1]),
                        label: `Price ( Past ${days} Days ) in AUD`,
                        borderColor: "#EEBC1D",
                      },
                    ],
                  }}
                />
              </div>
              <div className="coinChartButtons">
                {chartTimelines.map((timeline, index) => {
                  return (
                    <span key={index}>
                      <Button
                        className="chartButton"
                        onClick={() => {
                          createHistory(timeline.value);
                        }}
                        variant="outlined"
                      >
                        {timeline.label}
                      </Button>
                    </span>
                  );
                })}
              </div>
            </>
          )}
        </div>
      </ThemeProvider>
    </>
  );
};

export default CoinChart;
