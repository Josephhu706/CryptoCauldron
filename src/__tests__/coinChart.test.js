import {
  render,
  screen,
  cleanup,
  getByTestId,
  fireEvent,
  waitFor,
  getByText,
  act,
} from "@testing-library/react";
import CoinChart from "../components/CoinChart";
import "@testing-library/jest-dom";
import axiosMock from "axios";
import renderer from "react-test-renderer";
import { fetchCoin, fetchHistory } from "../config/API";

afterEach(() => {
  cleanup();
});

jest.mock("axios");
jest.mock("../config/API");

const mockedUsedNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUsedNavigate,
}));

test("Coin Chart heading should be rendered", async () => {
  await waitFor(() => {
    render(<CoinChart />);
    const heading = screen.getByText("Coin Price Data");
    expect(heading).toBeInTheDocument();
  });
});

test("Coin Chart should be rendered", async () => {
  axiosMock.get.mockResolvedValueOnce({
    data: [
      [1655189431100, 32260.558378485326],
      [1655189693519, 32374.969532405597],
    ],
  });
  act(() => render(<CoinChart />));
  const loading = screen.getByTestId("loading");
  expect(loading).toBeInTheDocument();
  expect(fetchHistory).toHaveBeenCalledTimes(1);
  waitFor(() => {
    const coinChart = screen.findByTestId("coinChart");
    expect(coinChart).toBeInTheDocument();
  });
});

test("Buttons should be rendered", async () => {
  axiosMock.get.mockResolvedValueOnce({
    data: [
      [1655189431100, 32260.558378485326],
      [1655189693519, 32374.969532405597],
    ],
  });
  act(() => render(<CoinChart />));
  const loading = screen.getByTestId("loading");
  expect(loading).toBeInTheDocument();
  expect(fetchHistory).toHaveBeenCalledTimes(1);
  waitFor(() => {
    const chartButton = screen.findByRole("button");
    expect(chartButton).toBeInTheDocument();
    const oneDay = screen.getByText("24 HOURS");
    expect(oneDay).toBeInTheDocument();
    const oneMonth = screen.getByText("30 DAYS");
    expect(oneMonth).toBeInTheDocument();
    const threeMonths = screen.getByText("3 MONTHS");
    expect(threeMonths).toBeInTheDocument();
    const oneYear = screen.getByText("1 YEAR");
    expect(oneYear).toBeInTheDocument();
  });
});

test("Render Error Message if Fetch Fails", async () => {
  axiosMock.get.mockRejectedValue({ data: {} });
  act(() => render(<CoinChart />));
  const loading = screen.getByTestId("loading");
  expect(loading).toBeInTheDocument();
  await waitFor(() => {
    const errorMsg = screen.getByTestId("fetchError");
    expect(errorMsg).toBeInTheDocument();
  });
});

test("Coin Chart matches snapshot", () => {
  const tree = renderer.create(<CoinChart />).toJSON();
  expect(tree).toMatchSnapshot();
});
