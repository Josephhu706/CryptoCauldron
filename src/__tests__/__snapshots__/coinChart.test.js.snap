// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Coin Chart matches snapshot 1`] = `
<div
  className="container"
>
  <h1
    className="coinHeading"
  >
    Coin Price Data
  </h1>
  <div
    className="coinChart"
    data-testid="coinChart"
  >
    <canvas
      height={150}
      role="img"
      width={300}
    />
  </div>
  <div
    className="coinChartButtons"
  >
    <span>
      <button
        className="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root chartButton css-1vbvugf-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        24 Hours
      </button>
    </span>
    <span>
      <button
        className="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root chartButton css-1vbvugf-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        30 Days
      </button>
    </span>
    <span>
      <button
        className="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root chartButton css-1vbvugf-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        3 Months
      </button>
    </span>
    <span>
      <button
        className="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root chartButton css-1vbvugf-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        1 Year
      </button>
    </span>
  </div>
</div>
`;
